### ISSUE TITLE
"offset" intrinsic should check the bounds, "arith_offset" must take element size into account

### ISSUE THREAD

**Issue by RalfJung** [solson/miri#183]
The following code (exercising the `offset` intrinsic) should not reach the panic because the out-of-bounds `offset` is UB, but currently, it does:
```rust
fn main() {
    let v = [0i8; 4];
    let x = &v as *const i8;
    let x = unsafe { x.offset(5) };
    panic!("this should never print: {:?}", x);
}
```

The following code (exercising the `arith_offset`) produces the wrong result -- it adds 1 to the pointer, instead of 1*size, i.e., 2:
```rust
fn main() {
    let v = [1i16, 2];
    let x = &v as *const i16;
    let x = x.wrapping_offset(1);
    assert_eq!(unsafe { *x }, 2);
}
```

**Comment by RalfJung** [comment]
In fixing this, I noticed that Pointer::offset does not actually do any provisions against overflowing. This makes it possible for a well-formed program to cause a panic in miri:
```rust
fn main() {
    let v = [1i16, 2];
    let x = &v[1] as *const i16;
    // Adding 2*isize::max and then 1 is like substracting 1
    let x = x.wrapping_offset(isize::max_value());
    let x = x.wrapping_offset(isize::max_value());
    let x = x.wrapping_offset(1);
    assert_eq!(unsafe { *x }, 1);
}
```
I feel like Pointer::offset (as well as signed_offset) should change their return type to involve EvalResult, and they should appropriately produce an error when something overflows. However, these methods are called a lot, so this will bubble up quite a bit. Still, I don't see a reasonable alternative -- do you?

**Comment by eddyb** [comment]
Shouldn't they just use `wrapping_add`?

**Comment by RalfJung** [comment]
I would assume the intrinsic and MIR operation have UB when things are overflowing? Anyway that's what I implemented now.

**Comment by eddyb** [comment]
Those 2 should do the in-bounds check which will fail for large val- oh I see hehe. You want `Pointer::overflowing_offset` using `overflowing_add` to allow the `wrapping_offset` overflows and disallow `offset` overflows.

**Comment by RalfJung** [comment]
Right. I pushed the current state of affairs to <https://github.com/RalfJung/miri/tree/pointers>.

**Comment by RalfJung** [comment]
Fixed by #185
